olive/model/config/hf_config.py

```diff
@@ -245,3 +246,11 @@ def task_or_model_class_required(cls, v, values):
             if not v and not values.get("task", None):
                 raise ValueError("Either task or model_class must be specified")
         return v
+
+
+def get_loading_args_from_pretrained(hf_config: HfConfig) -> Dict[str, Any]:
```

Comment:
>get_loading_args_from_pretrained [](http://example.com/codeflow?start=4&length=32)

better to make it be method of HfConfig class.

Reply:
Both functions members of `HFConfig` class?

Reply:
just  get_from_pretrained_args in HfConfig is enough.

Reply:
```
class HfConfig:
....
    def get_loading_args_from_pretrained(self) -> Dict[str, Any]:
        return self.from_pretrained_args.get_loading_args() if self.from_pretrained_args else {}

    def get_model_type(self) -> str:
        return get_hf_model_config(self.model_name, **self.get_loading_args_from_pretrained()).model_type
```

Reply:
```
class HfConfig:
     def get_from_pretrained_args(self) -> Dict[str, Any]:
        return self.from_pretrained_args.get_loading_args() if self.from_pretrained_args else {}
```

in hf_utils.py

```
def get_model_type(model_name: str, **kwargs):
    return get_hf_model_config(model_name, **kwargs).model_type
```

Reply:
Every calling site will have to duplicate code -

```
hf_config = HfConfig(**model_config["hf_config"])
model_type = get_model_type(hf_config.name, **hf_config.get_loading_args_from_pretrained())
```

Then why bother the function at all -

```
hf_config = HfConfig(**model_config["hf_config"])
model_type = get_hf_model_config(hf_config.name, **hf_config.get_loading_args_from_pretrained()).model_type
```

Reply:
I think it is ok since they are just call these functions/methods.

Reply:
by using the explicitly passing the name and args, we could decouple dependency of HfConfig and get_model_type.

Reply:
To avoid the circular dependency, HfConfig should be treated as a data object and shouldn't have any `smart` functions. All _logic_ based functions should be part of hf_utils which was my initial implementation. Adding functions in both hf_config and hf_utils is more likely to create circular dependency.

I think the restriction of not using HtConfig in hf_utils is limited. That should be relaxed, in fact, that should be encouraged.

The proposed `kwargs` makes code browsing hard since it doesn't explicitly say what these args are? Given that these args are actually *required* making them variable kwargs seems wrong.

Reply:
The decoupling is coming at the cost of code readability. My initial approach of adding the functions in hf_utils was better, in my opinion, since function arguments were explicit and hf_utils was above the `HfConfig` data class.

Reply:
What about this as a compromise https://github.com/microsoft/Olive/commit/cf60a73c9f2c0b8d769b604f3d0f04f839d9ca72? 

This way `olive.model.utils` is always the lowest level submodule under `olive.model` and we can avoid the circular import. utils should not import from eithe handler or config. That's how I personally arrange the hierarchy of utils submodules. 

We can also make this change later if you would like to merge your PR as is first.

Reply:
> The decoupling is coming at the cost of code readability. My initial approach of adding the functions in hf_utils was better, in my opinion, since function arguments were explicit and hf_utils was above the `HfConfig` data class.

Yes, you are correct. the cost is the code readabilities. Anyway, it is an open discussion. and I am ok to your changes and it is not a big issue if we import hf_utils in HfConfig.

Reply:
Resolving with introducing hf_utils dependency in hf_config

Reply:
Thanks!